geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/MessageDispatcher.java

```diff
@@ -362,31 +387,67 @@ protected void runDispatcher() {
           }
           waitForResumption();
         }
-        try {
-          clientMessage = (ClientMessage) _messageQueue.peek();
-        } catch (RegionDestroyedException skipped) {
-          break;
+
+        // if message is not delivered due to authentication expiation, this clientMessage
+        // would not be null.
+        if (clientMessage == null) {
+          try {
+            clientMessage = (ClientMessage) _messageQueue.peek();
+          } catch (RegionDestroyedException skipped) {
+            break;
+          }
         }
+
         getStatistics().setQueueSize(_messageQueue.size());
         if (isStopped()) {
           break;
         }
-        if (clientMessage != null) {
-          // Process the message
-          long start = getStatistics().startTime();
-          //// BUGFIX for BUG#38206 and BUG#37791
-          boolean isDispatched = dispatchMessage(clientMessage);
-          getStatistics().endMessage(start);
-          if (isDispatched) {
+
+        if (clientMessage == null) {
+          _messageQueue.remove();
+          continue;
+        }
+
+        // Process the message
+        long start = getStatistics().startTime();
+        try {
+          if (dispatchMessage(clientMessage)) {
+            getStatistics().endMessage(start);
             _messageQueue.remove();
             if (clientMessage instanceof ClientMarkerMessageImpl) {
               getProxy().setMarkerEnqueued(false);
             }
           }
-        } else {
+          clientMessage = null;
+          wait_for_re_auth_start_time = -1;
+        } catch (NotAuthorizedException notAuthorized) {
+          // behave as if the message is dispatched, remove from the queue
+          logger.info("skip delivering message: " + clientMessage, notAuthorized);
           _messageQueue.remove();
+          clientMessage = null;
+        } catch (AuthenticationExpiredException expired) {
+          if (wait_for_re_auth_start_time == -1) {
+            wait_for_re_auth_start_time = System.currentTimeMillis();
+            // only send the message to clients who can handle the message
+            if (getProxy().getVersion().isNewerThanOrEqualTo(RE_AUTHENTICATION_START_VERSION)) {
+              EventID eventId = createEventId();
+              sendMessageDirectly(new ClientReAuthenticateMessage(eventId));
+            }
+            // for older client, we still wait, just in case client will perform some operations to
+            // trigger credential refresh on its own.
+            Thread.sleep(200);
+          } else {
+            long elapsedTime = System.currentTimeMillis() - wait_for_re_auth_start_time;
+            if (elapsedTime > reAuthenticateWaitTime) {
+              logger.warn("Client did not re-authenticate back successfully in " + elapsedTime
+                  + "ms. Unregister this client proxy.");
+              pauseOrUnregisterProxy(expired);
+            } else {
+              Thread.sleep(200);
```

Comment:
I don't understand why the `sleep()` is needed here. Please explain.

Reply:
the `sleep()` is basically an interval between two authorization checks.  We don't want the check to be too busy. So the logic is like this
`
while(I am not stopped) {
   if (last message is not delivered), keep process the same message
   else take the message out of the queue
  try{
      dispatch the message (which includes authorization)
      message dispatched
  } catch (AuthExpirationException e) {
     if (this is the first time I am processing this message) {
          send the re-auth message to the client
          wait for 200ms before go back to the loop
     } else {
         if ( I waited  long enough) {
              stop the dispatcher and terminate the connection 
         } else {
             wait for 200ms and then go back to thee loop
         }
     }
  }
}
`

Reply:
Can the sleep be replaced with wait/notify?
Here it is expecting client to do a re-auth...Can a wait be done for re-auth to happen from client and then notify this wait when re-auth is received from client (from the re-auth command handling code).

Reply:
Thanks for that explanation. It helps a lot.

Whenever the dispatcher catches an `AuthExpirationException`, if it doesn't terminate the connection, all message processing is paused for 200ms.

I'm wondering how often does authentication expire generally? The product doesn't ship with a security manager that implements expiration (other than a test class), so it would be up to each particular customer's security manager implementation right?

Do you think it's ok to sleep for 200ms each time or do you think it would be worth the effort to shorten that time window by introducing say, a latch (we'd do an `await()` at the points where we're currently sleeping, and we'd issue a `countDown()` at the point where the client re-authenticates). **Is there an obvious place in the product where we could issue that `countDown()`?**

PS: In evaluating this change I looked for how `ClientReAuthenticateMessage` is used but I don't see it used anywhere in the product except where it's thrown in this class (and registered as a DSFID). That leads me to believe that a user's security manager would have to register a handler for that message. `ClientReAuthenticateMessage` is not mentioned in the product documentation nor in examples so I'm just guessing here.

Reply:
> Can the sleep be replaced with wait/notify?
> Here it is expecting client to do a re-auth...Can a wait be done for re-auth to happen from client and then notify this wait when re-auth is received from client (from the re-auth command handling code).

We are actually not waiting for client to re-auth, 'cause 1) the client might be an old client that will never re-authenticate back, 2) even the new client maybe not re-authenticate back successfully. As far as the server is concerned, I am just keep trying to "authorize" the message for a certain period of time (5 seconds by default), if in the end of this period, I still can not deliver this message, then I am closing this connection.

Reply:
> Thanks for that explanation. It helps a lot.
> 
> Whenever the dispatcher catches an `AuthExpirationException`, if it doesn't terminate the connection, all message processing is paused for 200ms.
> 
> I'm wondering how often does authentication expire generally? The product doesn't ship with a security manager that implements expiration (other than a test class), so it would be up to each particular customer's security manager implementation right?
> 
> Do you think it's ok to sleep for 200ms each time or do you think it would be worth the effort to shorten that time window by introducing say, a latch (we'd do an `await()` at the points where we're currently sleeping, and we'd issue a `countDown()` at the point where the client re-authenticates). **Is there an obvious place in the product where we could issue that `countDown()`?**
> 
> PS: In evaluating this change I looked for how `ClientReAuthenticateMessage` is used but I don't see it used anywhere in the product except where it's thrown in this class (and registered as a DSFID). That leads me to believe that a user's security manager would have to register a handler for that message. `ClientReAuthenticateMessage` is not mentioned in the product documentation nor in examples so I'm just guessing here.

Like I replied to Anil, there is not a definite event that the server needs to "wait" for,  the client may never re-authenticate back, so all the server can rely on is to keep trying to deliver the message with the "subject" it knows about (with 200 ms interval) for 5 seconds (configurable by system property).

Reply:
As we know sleep is not the best way to wait for certain state to be changed; its a last resort if there is no option. Based on the explanation, what I see is the code is expecting a state change from two different client:
1. Old client:
In this case, if it is an old client disconnect the server to client connection immediately. At this level we know the clients versions (from the client proxy); if it is an old version, the connection could be closed immediately without any wait.
2. New Client: 
In this case, the expectation is the new client will re-auth; for java clients the re-auth is processed by the command or security service; using the client info the wait in the dispatch could be notified. 
These are just thoughts, not sure if there are any hurdle in doing this way. It will be good investigate to see if this is possible.

Reply:
Sure， we can add a task to investigate this, but I don't want to further muddle up this PR anymore.

Reply:
About the client, we don't immediately close the connection to old client and still choose to wait because any active client interaction will trigger a re-auth automatically, so there is still chance that the message could be delivered to the old client, we have a test specifically for that: `AuthExpirationDUnitTest.cqOlderClientWithClientInteractionWillDeliverEventEventually`

Reply:
Adding wait/notify semantics between the client re-authenticating and the `MessageDispatcher` waiting to start trying (to dispatch) again will add some complexity to the existing `SecurityManager` framework.

I now see that this PR (and associated ticket) are one task in a larger effort [GEODE-9451](https://issues.apache.org/jira/browse/GEODE-9451) covered by an [RFC](https://cwiki.apache.org/confluence/display/GEODE/On+Demand+Geode+Authentication+Expiration+and+Re-authentication).

In that context, I think @jinmeiliao's answer: "we can add a task to investigate this, but I don't want to further muddle up this PR anymore" is acceptable.